optimism/src/keccak/mod.rs

```diff
@@ -70,6 +79,77 @@ pub enum Constraint {
     IotaStateG(usize),
 }
 
+#[allow(dead_code)]
+/// The Keccak circuit
+pub(crate) type KeccakCircuit<F> = Circuit<ZKVM_KECCAK_COLS, Steps, F>;
+
+#[allow(dead_code)]
+impl<F: Field> KeccakCircuit<F> {
+    /// Create a new Keccak circuit
+    pub fn new(domain_size: usize) -> Self {
```

Comment:
When you first create a keccak circuit, it starts with an empty witness with capacity the domain size. And then the lookups and constraints, which can be known ahead of time, are stored in the corresponding fields.